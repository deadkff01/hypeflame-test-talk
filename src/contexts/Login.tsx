import React, { useContext, useReducer } from "react";

type ContextProps = {
  isLogged: boolean;
};

type LoginContextProps = {
  state: ContextProps;
  dispatch: (a: Action) => void;
};

export enum ACTIONS {
  LOGIN = "LOGIN",
  LOGOUT = "LOGOUT",
}

export type Action = {
  type: ACTIONS;
  payload: boolean;
};

interface ILoginProvider {
  children: React.ReactNode;
}

const initialState: ContextProps = {
  isLogged: false,
};

export const LoginContext = React.createContext<LoginContextProps>({
  state: initialState,
  dispatch: () => {},
});

export const reducer = (state: ContextProps, action: Action) => {
  const { type, payload } = action;
  switch (type) {
    case ACTIONS.LOGIN: {
      return { ...state, isLogged: payload };
    }
    case ACTIONS.LOGOUT: {
      return { ...state, isLogged: false };
    }
    default:
      return state;
  }
};

export const LoginProvider = ({ children }: ILoginProvider) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const store = { state, dispatch };
  return (
    <LoginContext.Provider value={store}>{children}</LoginContext.Provider>
  );
};

export const LoginConsumer = LoginContext.Consumer;

export const useLoginContext = () => useContext(LoginContext);
