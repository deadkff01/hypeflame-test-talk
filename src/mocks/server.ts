import { rest } from "msw";
import { setupServer } from "msw/node";

const responseMock = {
  username: "hypeflame",
};

export const server = setupServer(
  rest.post(`${process.env.REACT_APP_BACKEND}/login`, (_, res, ctx) => {
    return res(ctx.json(responseMock));
  })
);

export { rest };
