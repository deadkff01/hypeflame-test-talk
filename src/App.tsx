import GlobalStyles from "components/GlobalStyles";
import "./App.css";
import { Routes } from "Routes";
import { LoginProvider, LoginConsumer } from "./contexts/Login";

function App() {
  return (
    <LoginProvider>
      <LoginConsumer>
        {({ state }) => (
          <>
            <GlobalStyles />
            <Routes isLogged={state?.isLogged} />
          </>
        )}
      </LoginConsumer>
    </LoginProvider>
  );
}

export default App;
