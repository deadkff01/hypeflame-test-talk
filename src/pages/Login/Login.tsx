import { useState } from "react";
import { RouteComponentProps } from "@reach/router";
import { tw } from "twind";
import Container from "components/Container";
import Input from "components/Input";
import { useLoginContext, ACTIONS } from "contexts/Login";
import { navigate } from "@reach/router";
import { loginRequest } from "services/login";
import Button from "components/Button";
import Loader from "components/Loader";

export const Login: React.FC<RouteComponentProps> = () => {
  const { dispatch } = useLoginContext();
  const [username, setUsername] = useState<string>("");
  const [usernameError, setUsernameError] = useState<boolean>(false);
  const [password, setPassword] = useState<string>("");
  const [passwordError, setPasswordError] = useState<boolean>(false);
  const [loginError, setLoginError] = useState<boolean>(false);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const blurUsername = () => setUsernameError(username.length < 3);

  const blurPassword = () => setPasswordError(password.length < 6);

  const btnLoginDisabled = username.length < 3 || password.length < 6;

  const doLogin = async () => {
    try {
      setIsLoading(true);
      const body = {
        username,
        password,
      };
      await loginRequest(body);
      dispatch({ type: ACTIONS.LOGIN, payload: true });
      navigate("/user-logged");
    } catch {
      setLoginError(true);
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <Container>
      <div>
        <h1 className={tw`font-bold text-4xl my-5`}>Login</h1>
        <div className={tw`flex justify-center px-3`}>
          <div className={tw`w-full md:w-6/12 lg:w-4/12`}>
            <Input
              value={username}
              placeholder="Username"
              id="username"
              onBlur={blurUsername}
              error={usernameError}
              errorMessage={"O Username deve ter no mínimo 3 caracteres"}
              onChange={({ target }: React.ChangeEvent<HTMLInputElement>) => {
                setUsername(target.value);
                setUsernameError(false);
                setLoginError(false);
              }}
            />
            <Input
              value={password}
              placeholder="Password"
              id="password"
              type="password"
              error={passwordError}
              errorMessage={"O Password deve ter no mínimo 6 caracteres"}
              onBlur={blurPassword}
              onChange={({ target }: React.ChangeEvent<HTMLInputElement>) => {
                setPassword(target.value);
                setPasswordError(false);
                setLoginError(false);
              }}
            />
          </div>
        </div>
        {loginError ? (
          <div className={tw`font-bold my-2 text-red-500`}>
            Erro ao efetuar o login!
          </div>
        ) : null}
        <Button
          id="btn-login"
          data-testid="btn-login"
          className={tw`mt-2`}
          disabled={btnLoginDisabled}
          onClick={doLogin}
        >
          Entrar
        </Button>
      </div>

      {isLoading ? <Loader /> : null}
    </Container>
  );
};
