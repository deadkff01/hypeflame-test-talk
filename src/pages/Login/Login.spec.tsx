import { screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithRouter } from "../../testUtils";
import { Login } from "./Login";
import { LoginContext, LoginConsumer } from "contexts/Login";
import { server, rest } from "mocks/server";
import App from "../../App";

const dispatch = () => {};
const state = {
  isLogged: false,
};

const store = { state, dispatch };

describe("Test Login", () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  test("Snapshot", () => {
    const { container } = renderWithRouter(
      <LoginContext.Provider value={store}>
        <LoginConsumer>{() => <Login />}</LoginConsumer>
      </LoginContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test("fill inputs success", async () => {
    renderWithRouter(
      <LoginContext.Provider value={store}>
        <LoginConsumer>{() => <Login />}</LoginConsumer>
      </LoginContext.Provider>
    );

    const username = await screen.findByTestId("username");
    userEvent.type(username, "hypeflame");

    const password = await screen.findByTestId("password");
    userEvent.type(password, "test123");

    const btnLogin = await screen.findByTestId("btn-login");

    expect(btnLogin).not.toBeDisabled();
  });

  test("fill inputs error", async () => {
    renderWithRouter(
      <LoginContext.Provider value={store}>
        <LoginConsumer>{() => <Login />}</LoginConsumer>
      </LoginContext.Provider>
    );

    const username = await screen.findByTestId("username");
    userEvent.type(username, "hy");

    const password = await screen.findByTestId("password");
    userEvent.type(password, "test");

    userEvent.click(username);

    const usernameError = await screen.findByText(
      "O Username deve ter no mínimo 3 caracteres"
    );

    const passwordError = await screen.findByText(
      "O Password deve ter no mínimo 6 caracteres"
    );

    expect(usernameError).toBeDefined();

    expect(passwordError).toBeDefined();
  });

  test("login success", async () => {
    const {
      history: { navigate },
    } = renderWithRouter(<App />);

    const username = await screen.findByTestId("username");
    userEvent.type(username, "hypeflame");

    const password = await screen.findByTestId("password");
    userEvent.type(password, "test123");

    const btnLogin = await screen.findByTestId("btn-login");
    userEvent.click(btnLogin);

    await act(async () => {
      await navigate("/user-logged");
    });

    const userLogged = await screen.findByText("Usuário logado");
    expect(userLogged).toBeDefined();
  });

  test("login error", async () => {
    server.use(
      rest.post(`${process.env.REACT_APP_BACKEND}/login`, (_, res, ctx) => {
        return res(ctx.status(400));
      })
    );

    renderWithRouter(
      <LoginContext.Provider value={store}>
        <LoginConsumer>{() => <Login />}</LoginConsumer>
      </LoginContext.Provider>
    );

    const username = await screen.findByTestId("username");
    userEvent.type(username, "hypeflame");

    const password = await screen.findByTestId("password");
    userEvent.type(password, "test123");

    const btnLogin = await screen.findByTestId("btn-login");
    userEvent.click(btnLogin);

    const errorMessage = await screen.findByText("Erro ao efetuar o login!");
    expect(errorMessage).toBeDefined();
  });
});
