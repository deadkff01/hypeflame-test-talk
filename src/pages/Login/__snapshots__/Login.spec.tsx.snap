// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Login Snapshot 1`] = `
<div>
  <div
    class="sc-bdnxRM hBgoUI"
  >
    <div>
      <h1
        class="font-bold text-4xl my-5"
      >
        Login
      </h1>
      <div
        class="flex justify-center"
      >
        <div
          class="w-3/12"
        >
          <div
            class="sc-gtsrHT dnQHDr"
          >
            <div
              class="control"
            >
              <input
                data-testid="username"
                id="username"
                type="text"
                value=""
              />
              <label
                class=""
                for="username"
              >
                Username
              </label>
            </div>
            <span
              class="error "
              data-testid="error-username"
            />
          </div>
          <div
            class="sc-gtsrHT dnQHDr"
          >
            <div
              class="control"
            >
              <input
                data-testid="password"
                id="password"
                type="password"
                value=""
              />
              <label
                class=""
                for="password"
              >
                Password
              </label>
            </div>
            <span
              class="error "
              data-testid="error-password"
            />
          </div>
        </div>
      </div>
      <button
        class="sc-dlnjwi iFxcWs mt-2"
        data-testid="btn-login"
        disabled=""
        id="btn-login"
      >
        Entrar
      </button>
    </div>
  </div>
</div>
`;
